DOCS-7747, DOCS-7795 Adds APM Deployment Tracking Documentation (#22866)
DataDog/documentation: content/en/dora_metrics/deployments/_index.md

further_reading:
- link: "/dora_metrics/failures"
  tag: "Documentation"
  text: "Learn about sending failure events"
- link: "/tracing/service_catalog"
  tag: "Documentation"
  text: "Learn about the Service Catalog"
- link: "https://github.com/DataDog/datadog-ci"
  tag: "Source Code"
  text: "Learn about the datadog-ci CLI tool"
- link: "/continuous_delivery/deployments"
  tag: "Documentation"
  text: "Learn about Deployment Visibility"
---

{{< site-region region="gov" >}}
<div class="alert alert-warning">DORA Metrics is not available in the selected site ({{< region-param key="dd_site_name" >}}) at this time.</div>
{{< /site-region >}}

## Overview

Deployment events are used to compute [deployment frequency](#calculating-deployment-frequency), [change lead time](#calculating-change-lead-time), and [change failure rate](#calculating-change-failure-rate). 

## Selecting a deployment data source

{{< whatsnext desc="DORA Metrics supports the following data sources for deployment events. See the respective documentation to set up the data source for your deployment events:" >}}
  {{< nextlink href="/dora_metrics/deployments/apm" >}}APM Deployment Tracking{{< /nextlink >}}

>   {{< nextlink href="/dora_metrics/deployments/" >}}Deployment Event API or use the datadog-ci CLI{{< /nextlink >}}

Reviewer: ```suggestion
  {{< nextlink href="/dora_metrics/deployments/deployment_api" >}}Deployment Event API or use the datadog-ci CLI{{< /nextlink >}}
```

## Overview

> To send your own deployment events, use the [DORA Metrics API][1] or the [`datadog-ci dora deployment`][2] command. The following attributes are required:

Reviewer 2: should we mention in the overview that users can either send deployments via the API or APM Deployments tracking?

  {{< nextlink href="/dora_metrics/deployments/apm" >}}APM Deployment Tracking{{< /nextlink >}}
  {{< nextlink href="/dora_metrics/deployments/" >}}Deployment Event API or use the datadog-ci CLI{{< /nextlink >}}
{{< /whatsnext >}}

## Calculating deployment frequency

Deployment frequency is calculated based on the `dora.deployments.count` metric that is generated and increased with each deployment detected from your selected deployment data source. Frequency is calculated by dividing `dora.deployments.count` over a specific time frame.

## Calculating change lead time

For a single Git commit, the change lead time (CLT) is calculated as time from the creation of the commit to when the deployment including that commit was executed.

To calculate the change lead time for a deployment, Datadog runs [`git log`][6] between the deployment commit SHA and the previous deployment commit SHA to find all the commits being deployed. Then, it computes the average of the related change lead time values. Datadog doesn't store the actual content of files in your repository, only Git commit and tree objects.

> There are two requirements for calculating change lead time:

Reviewer: I think we can remove this in favour of specifying everything needed in the different deployment sources pages. For example, point 1 is only important for API deployments. 

I would include in the APM deployments and API deployments a very similar sentence/setup, with the repository metadata synchronization which is in common (can both be duplicated or we could add a different page for that) and additional setup based on the source:

1. For API, deployments need to contain specific attributes
2. For APM, telemetry should be tagged with Git information

Author: IMO it makes sense to have a section per metric in this page so when the user skims it, they can tell from the headings which metrics deployment data is relevant for. Each section can provide brief summary for how deployments are used to calculate the metrics, and individual details can be left for the APM and API pages

1. Both the Git repository URL and commit SHA are provided when sending deployment events.
2. Your repository metadata is being [synchronized to Datadog](#synchronize-repository-metadata-to-datadog).

### Breakdown metrics

Datadog breaks down change lead time into the following metrics, which represent the different stages from commit creation to deployment.

> To compute these metrics, there must be a pull request (PR) associated with a commit, if any. A commit is associated with a PR if the commit is first introduced to the target branch when merging that PR.

Reviewer: I think we should mention somewhere that the breakdown metrics (at least those related to PRs), are only created when the source of the repository metadata is GitHub

Reviewer 3: ```suggestion
These metrics are only computed when the source of the repository metadata is GitHub, and there must be a pull request (PR) associated with a commit, if any. A commit is associated with a PR if the commit is first introduced to the target branch when merging that PR.
```

If a commit does not have an associated PR, only time to deploy and deploy time metrics are available.

> - `dora.time_to_pr_ready`: Time from when the commit is created until the PR is ready for review. This metric is only available for commits that were made before the PR is ready for review.

Reviewer: ```suggestion
- `dora.time_to_pr_ready`: Time from when the commit is created until the PR is ready for review. This metric is only available for commits that were made before the PR was marked as ready for review.
```

> ### Breakdown metrics

Reviewer 2: should we move Breakdown metrics to content/en/dora_metrics/data_collected/_index.md?

Reviewer: I agree!

Reviewer 3: Will do this in a follow up commit.
